src/openai/_tracing.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import json
+import functools
+from typing import TypeVar, Callable, Any, Generator, Iterator, Union, Literal
+from typing_extensions import ParamSpec
+
+from ._extras import opentelemetry as trace, has_tracing_enabled
+from .types.chat import ChatCompletion, ChatCompletionChunk
+from .types.completion import Completion
+from ._streaming import Stream
+
+TracedMessageTypes = Literal["system", "user", "assistant", "tool"]
+TracedModels = Union[ChatCompletion, Completion]
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+
+def _set_attribute(span: trace.Span, key: str, value: Any) -> None:
+    if value:
+        span.set_attribute(key, value)
+
+
+def _add_request_chat_message_event(span: trace.Span, **kwargs: Any) -> None:
+    for message in kwargs.get("messages", []):
+        try:
+            message = message.to_dict()
+        except AttributeError:
+            pass
+        
+        if message.get("role") in TracedMessageTypes:
+            name = f"gen_ai.{message.get('role')}.message"
+            span.add_event(
+                name=name,
+                attributes={"event.data": json.dumps(message, indent=2)}
+            )
+
+
+def _add_request_chat_attributes(span: trace.Span, **kwargs: Any) -> None:
+    _set_attribute(span, "gen_ai.system", "openai")
+    _set_attribute(span, "gen_ai.request.model", kwargs.get("model"))
+    _set_attribute(span, "gen_ai.request.max_tokens", kwargs.get("max_tokens"))
+    _set_attribute(span, "gen_ai.request.temperature", kwargs.get("temperature"))
+    _set_attribute(span, "gen_ai.request.top_p", kwargs.get("top_p"))
+
+
+def _add_response_chat_message_event(span: trace.Span, result: ChatCompletion) -> None:
+    for choice in result.choices:
+        response = {
+            "message.role": choice.message.role,
+            "message.content": choice.message.content,
+            "finish_reason": choice.finish_reason,
+            "index": choice.index,
+        }
+        if choice.message.tool_calls:
+            response.update({"message.tool_calls": "".join([tool.to_json() for tool in choice.message.tool_calls])})
```

Comment:
Don't we want some separator other than empty string? Also, this is just a single assignment, right?

```suggestion
            response["message.tool_calls"] = "".join([tool.to_json() for tool in choice.message.tool_calls])
```

Reply:
Good point, let me double-check on this